test/index.ts

```diff
@@ -154,3 +154,177 @@ test('resolveFromGit() with commit from non-github repo', async t => {
+    resolvedVia: 'git-repository',
+  })
+  t.end()
+})
+
+test('resolveFromGit() gitlab with branch', async (t) => {
+  const resolveResult = await resolveFromGit({pref: 'gitlab:huochunpeng/debug-npm#branch'})
+  t.deepEqual(resolveResult, {
+    id: 'gitlab.com/huochunpeng/debug-npm/f34a78790bc232e678c9169e5c30cba72458639b',
+    normalizedPref: 'gitlab:huochunpeng/debug-npm#branch',
+    resolution: {
+      tarball: 'https://gitlab.com/huochunpeng/debug-npm/repository/archive.tar.gz?ref=f34a78790bc232e678c9169e5c30cba72458639b',
+    },
+    resolvedVia: 'git-repository',
+  })
+  t.end()
+})
+
+test('resolveFromGit() gitlab with tag', async (t) => {
+  const resolveResult = await resolveFromGit({pref: 'gitlab:huochunpeng/debug-npm#v1.0.0'})
+  t.deepEqual(resolveResult, {
+    id: 'gitlab.com/huochunpeng/debug-npm/361ff261e60f60a91da09bb259230303ccef8087',
+    normalizedPref: 'gitlab:huochunpeng/debug-npm#v1.0.0',
+    resolution: {
+      tarball: 'https://gitlab.com/huochunpeng/debug-npm/repository/archive.tar.gz?ref=361ff261e60f60a91da09bb259230303ccef8087',
+    },
+    resolvedVia: 'git-repository',
+  })
+  t.end()
+})
+test('resolveFromGit() normalizes full url', async (t) => {
+  const resolveResult = await resolveFromGit({ pref: 'git+ssh://git@github.com:zkochan/is-negative.git#2.0.1' })
+  t.deepEqual(resolveResult, {
+    id: 'github.com/zkochan/is-negative/2fa0531ab04e300a24ef4fd7fb3a280eccb7ccc5',
+    normalizedPref: 'github:zkochan/is-negative#2.0.1',
+    resolution: {
+      tarball: 'https://codeload.github.com/zkochan/is-negative/tar.gz/2fa0531ab04e300a24ef4fd7fb3a280eccb7ccc5',
+    },
+    resolvedVia: 'git-repository',
+  })
+  t.end()
+});
+
+test('resolveFromGit() normalizes full url (alternative form)', async (t) => {
+  const resolveResult = await resolveFromGit({pref: 'git+ssh://git@github.com/zkochan/is-negative.git#2.0.1'})
+  t.deepEqual(resolveResult, {
+    id: 'github.com/zkochan/is-negative/2fa0531ab04e300a24ef4fd7fb3a280eccb7ccc5',
+    normalizedPref: 'github:zkochan/is-negative#2.0.1',
+    resolution: {
+      tarball: 'https://codeload.github.com/zkochan/is-negative/tar.gz/2fa0531ab04e300a24ef4fd7fb3a280eccb7ccc5',
+    },
+    resolvedVia: 'git-repository',
+  })
+  t.end()
+});
+
+test('resolveFromGit() normalizes full url (alternative form 2)', async (t) => {
+  const resolveResult = await resolveFromGit({ pref: 'https://github.com/zkochan/is-negative.git#2.0.1' })
+  t.deepEqual(resolveResult, {
+    id: 'github.com/zkochan/is-negative/2fa0531ab04e300a24ef4fd7fb3a280eccb7ccc5',
+    normalizedPref: 'github:zkochan/is-negative#2.0.1',
+    resolution: {
+      tarball: 'https://codeload.github.com/zkochan/is-negative/tar.gz/2fa0531ab04e300a24ef4fd7fb3a280eccb7ccc5',
+    },
+    resolvedVia: 'git-repository',
+  })
+  t.end()
+});
+
+// This test relies on implementation detail.
+// current implementation does not try git ls-remote --refs on pref with full commit hash, this fake repo url will pass.
```

Comment:
so what makes it private then?

Reply:
Not this one, the one after. The fake repo fails check on both git url and https url, effectively runs the same code path of private repo by using ssh url.